src/color-extractor.ts

```diff
@@ -0,0 +1,364 @@
+const { rgb2lab, deltaE } = require('rgb-lab');
+
+
+/**
+ * Takes in an ImageData object and finds the k most dominant colors from
+ * the image data. Also returns the number of pixels assoiated with each 
+ * color's cluster. Dominant colors that are too close to colors in 
+ * colorsToIgnore are ignored.
+ * 
+ * @param myData          ImageData generated from a HTML Canvas element
+ * @param numClusters     Number of dominant colors to find
+ * @param colorsToIgnore  Array of colors as RGB strings to ignore
+ * @param threshold       Threshold for similarity in LAB color space
+ *                        for ignoring dominant color
+ */
+export function getDominantColors(myData: ImageData, numClusters: number,
+  colorsToIgnore: Array<string>, threshold: number): Array<any> {
+  console.log(colorsToIgnore);
+  let dataMap, keyArray, total;
+  [dataMap, keyArray, total] = getDataMap(myData);
+
+  let centerCandidates = getCenterCandidates(dataMap, keyArray, total);
+
+  let centers = new Array<string>();
+  let centerCounts, centerToColor;
+  while (centers.length === 0) {
+    [centers, centerCounts, centerToColor] = kMeans(numClusters, dataMap, centerCandidates);
+    console.log(centers);
+    [centers, centerCounts] = filterColors(centers, centerCounts, centerToColor, 
+      colorsToIgnore, threshold);
+    console.log(centers);
+  }
+
+  return [centers, centerCounts];
+}
+
+/**
+ * Gets center candidates for performing k means on the data in dataMap.
+ * Eliminates candidates if the count for a certain color is too low.
+ * 
+ * @param dataMap   Map from color to count of color in image.
+ * @param keyArray  Array containing keys of dataMap in arbitrary order.
+ * @param total     Total number of pixels from image used to make dataMap.
+ */
+function getCenterCandidates(dataMap: Map<string, number>, 
+  keyArray: Array<string>, total: number) {
+  let centerCandidates = [];
+  const threshold = total / keyArray.length / 2 ;
+  for (let i = 0; i < keyArray.length; i++) {
+    const count = dataMap.get(keyArray[i]);
+    if (count && count > threshold){
+      centerCandidates.push(keyArray[i]);
+    }
+  }
+  return centerCandidates;
+}
+
+/**
+ * Filters out cluster centers that are too similar to the specified color.
+ * Also converts center color to nearest color found in the image.
+ * Returns array of valid centers and center counts for those centers.
+ * 
+ * @param centers         Array of colors that are centers.
+ * @param centerCounts    Map from center to count of number of pixels.
+ * @param centerToColor   Map from center to colors in center's cluster.
+ */
+function filterColors(centers: Array<string>,
+  centerCounts: Array<number>,
+  centerToColor: Map<string, Array<string>>,
+  colorsToIgnore: Array<string>, threshold: number): Array<any> {
+  console.log(colorsToIgnore);
+  for (let i = centers.length - 1; i >= 0; i--) {
```

Comment:
They should always be the same. Since the implementation is now object-oriented, I don't think someone calling the public functions will be able to change the size of centers/centerCounts.